### PR TITLE
Improve and extract internals of vcpkg_find_acquire_program.

### PR DESCRIPTION
**Pull request by BillyONeal**
This work is extracted from the tool ports work https://github.com/microsoft/vcpkg/pull/31748/ . Components of this function are going to be necessary for any tool ports future, so this extracts them from that PR to make it easier to review.

Overall changes to vcpkg_find_acquire_program and its protocols:

* vcpkg_extract_archive is extracted from vcpkg_find_acquire_program, and also used for
 vcpkg_extract_source_archive. In the future it might make sense to power this with vcpkg x-extract.
* vcpkg_download_sourceforge is extracted from vcpkg_from_sourceforge, which allows
 vcpkg_find_acquire_program to download from sourceforge without needing to break --editable
 in the process. This also reduces the divergence between sourceforge and normal downloads.
* The separate boolean supported_on_unix was removed in favor of testing whether means of
 acquiring the entity in the first place are known.

### CODE REVIEW

File: scripts/cmake/vcpkg_extract_source_archive.cmake [hunk 1]
```diff
@@ -2,13 +2,7 @@ function(z_vcpkg_extract_source_archive_deprecated_mode archive working_director
...
     if(NOT EXISTS "${working_directory}/${archive_filename}.extracted")
         message(STATUS "Extracting source ${archive}")
-        file(MAKE_DIRECTORY "${working_directory}")
-        vcpkg_execute_required_process(
-            ALLOW_IN_DOWNLOAD_MODE
-            COMMAND "${CMAKE_COMMAND}" -E tar xjf "${archive}"
-            WORKING_DIRECTORY "${working_directory}"
-            LOGNAME extract
-        )
+        vcpkg_extract_archive(ARCHIVE "${archive}" DESTINATION "${working_directory}")
```
Comment (BillyONeal) [inline scripts/cmake/vcpkg_extract_source_archive.cmake:5]: Note that the log file is intentionally removed. https://github.com/microsoft/vcpkg/pull/31748#discussion_r1212670551

**Comment by BillyONeal** [comment]
Tagged vcpkg-bug because this fixes `--editable` for any ports that want doxygen or swig.

Tagged vcpkg-feature because this adds new helper functions.

File: scripts/cmake/vcpkg_find_acquire_program(NUGET).cmake [hunk 2]
```diff
@@ -1,9 +1,11 @@
 set(program_name nuget)
 set(tool_subdirectory "5.11.0")
-set(paths_to_search "${DOWNLOADS}/tools/nuget-${tool_subdirectory}-windows")
 set(brew_package_name "nuget")
-set(download_urls "https://dist.nuget.org/win-x86-commandline/v5.11.0/nuget.exe")
-set(rename_binary_to "nuget.exe")
-set(download_filename "nuget.5.11.0.exe")
 set(raw_executable ON)
-set(download_sha512 06a337c9404dec392709834ef2cdbdce611e104b510ef40201849595d46d242151749aef65bc2d7ce5ade9ebfda83b64c03ce14c8f35ca9957a17a8c02b8c4b7)
+if(CMAKE_HOST_WIN32)
```
Comment (ras0219-msft) [inline scripts/cmake/vcpkg_find_acquire_program(NUGET).cmake:5]: nuget works on non-windows platforms via mono.

**Review by ras0219-msft — DISMISSED** [review]

File: scripts/cmake/vcpkg_find_acquire_program(NUGET).cmake (hunk 2 above)
Comment (BillyONeal) [reply scripts/cmake/vcpkg_find_acquire_program(NUGET).cmake:5]: That is true, but `vcpkg_find_acquire_program` does not expose a way to use it that way. (It's almost like implementing this in 3 different places is ... bad ? :) )

Comment (BillyONeal) [reply scripts/cmake/vcpkg_find_acquire_program(NUGET).cmake:5]: I guess there is `brew package name` :shrug:

File: scripts/cmake/vcpkg_find_acquire_program(JOM).cmake [hunk 3]
```diff
@@ -2,10 +2,12 @@ set(program_name jom)
 set(program_version_string 1_1_4)
 set(tool_subdirectory "jom-${program_version_string}")
 set(paths_to_search "${DOWNLOADS}/tools/jom/${tool_subdirectory}")
```
[the remark this answers is not in the document]
Comment (dg0yt) [reply scripts/cmake/vcpkg_find_acquire_program(JOM).cmake]: Can we generally use `CMAKE_HOST_WIN32`?
Cf. https://github.com/microsoft/vcpkg/pull/34546 for why.

File: scripts/cmake/vcpkg_extract_archive.cmake [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+            message(FATAL_ERROR "${arg_name} is required.")
+        endif()
+    endforeach()
+
+    if(DEFINED arg_UNPARSED_ARGUMENTS)
+        message(FATAL_ERROR "Unrecognized arguments: ${arg_UNPARSED_ARGUMENTS}")
+    endif()
+
+    if(EXISTS "${arg_DESTINATION}")
+        message(FATAL_ERROR "${arg_DESTINATION} was an extraction target, but it already exists.")
```
[the remark this answers is not in the document]
Comment (autoantwort) [reply scripts/cmake/vcpkg_extract_archive.cmake:19]: Yeah issues like https://github.com/microsoft/vcpkg/issues/36709 are popping up a lot recently

**Pull request merged**